projects/style_gen/modules.py

```diff
@@ -57,3 +71,319 @@ def get_temp_history(self, observation: Message) -> Optional[str]:
+            embedding=embedding,
+            dropout=opt['dropout'],
+            attention_dropout=opt['attention_dropout'],
+            relu_dropout=opt['relu_dropout'],
+            padding_idx=padding_idx,
+            learn_positional_embeddings=opt['learn_positional_embeddings'],
+            embeddings_scale=opt['embeddings_scale'],
+            n_positions=n_positions,
+            activation=opt['activation'],
+            variant=opt['variant'],
+            n_segments=n_segments,
+        )
+
+    def __init__(self, opt, dictionary, num_classes: int, personality_as_label: bool):
+        super().__init__(opt, dictionary)
+        self.classifier_head = nn.Linear(opt['embedding_size'], num_classes)
+        self.personality_as_label = personality_as_label
+
+    def forward(self, *xs):
+        """
+        Get output class logits from the model.
+
+        :param xs:
+            - list of inputs to the encoder/decoder. Elements:
+              - text_vec: (LongTensor[bsz, text seqlen])
+              - label_vec: (LongTensor[bsz, label seqlen])
+                  (Only used if not self.personality_as_label)
+        :return:
+            - the model's predicted per-class scores.
+              (FloatTensor[bsz, len(class_list)])
+        """
+
+        if self.personality_as_label:
+            # All tokens go into the encoder and classification is learned from that.
+            # This is useful in the standard case where we have a fixed utterance that
+            # doesn't need to be generated, and we can just stick it all in the encoder
+            # to be classified.
+            assert len(xs) == 1
+            # Only one input allowed
+            bsz = xs[0].size(0)
+            encoder_states = self.encoder(*xs)
+            inputs = self.START.detach().expand(bsz, 1)
+            # Generate most likely class given start token as input
+            latent, _ = self.decoder(inputs, encoder_states)
+            # latent: [bsz, seqlen, emb_dim]
+            scores = self.classifier_head(latent.squeeze(dim=1))
+        else:
+            # Tokens are split between the encoder and decoder and classification is
+            # learned from both. This is useful when we want to classify a partially
+            # generated utterance, along with its context in the encoder.
+            text_vec, label_vec = xs
+            encoder_states = self.encoder(text_vec)
+            latent, _ = self.decoder(label_vec, encoder_states)
+            # latent: [bsz, seqlen, emb_dim]
+            scores = self.classifier_head(latent.mean(dim=1))
+
+        return scores
+
+
+class BatchWithPersonalities(AttrDict):
+    """
+    Adds a 'personalities' field to the batch in the case where personality information
+    is not encoded in any other field.
+    """
+
+    def __init__(self, personalities=None, **kwargs):
+        super().__init__(personalities=personalities, **kwargs)
+
+
+class TransformerDecoderWithEmbeds(TransformerDecoder):
```

Comment:
This solution for passing in an `embedded_input` definitely works; another way you can get around this is by just subclassing `forward_embedding` to return the `input` directly (i.e. bypass the actual `embedding` phase); this would allow you to not have to repeat some of the decoder forward logic

Reply:
I wrote a few notes below about how some of the logic has been factored out but since this is a project agent it's probably good (if not better) to keep as is for maximum reproducibility (as those functions may change in the future...)

Reply:
> This solution for passing in an `embedded_input` definitely works; another way you can get around this is by just subclassing `forward_embedding` to return the `input` directly (i.e. bypass the actual `embedding` phase); this would allow you to not have to repeat some of the decoder forward logic

Yeah good point! I think this code pre-dates some of that more recent refactoring